lib/services/local/externalcloudaudit.go

```diff
@@ -0,0 +1,158 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package local
+
+import (
+	"context"
+	"time"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/types/externalcloudaudit"
+	"github.com/gravitational/teleport/api/types/header"
+	"github.com/gravitational/teleport/lib/backend"
+	"github.com/gravitational/teleport/lib/services"
+)
+
+const (
+	externalCloudAuditPrefix      = "external_cloud_audit"
+	externalCloudAuditDraftName   = "draft"
+	externalCloudAuditClusterName = "cluster"
+	externalCloudAuditLockName    = "external_cloud_audit_lock"
+	externalCloudAuditLockTTL     = 10 * time.Second
+)
+
+var (
+	draftExternalCloudAuditBackendKey   = backend.Key(externalCloudAuditPrefix, externalCloudAuditDraftName)
+	clusterExternalCloudAuditBackendKey = backend.Key(externalCloudAuditPrefix, externalCloudAuditClusterName)
+)
+
+// ExternalCloudAuditService manages external cloud audit resources in the Backend.
+type ExternalCloudAuditService struct {
+	backend backend.Backend
+}
+
+// NewExternalCloudAuditService creates a new ExternalAuditService.
+func NewExternalCloudAuditService(backend backend.Backend) (*ExternalCloudAuditService, error) {
+	return &ExternalCloudAuditService{
+		backend: backend,
+	}, nil
+}
+
+// GetDraftExternalCloudAudit returns the draft external cloud audit resource.
+func (s *ExternalCloudAuditService) GetDraftExternalCloudAudit(ctx context.Context) (*externalcloudaudit.ExternalCloudAudit, error) {
+	item, err := s.backend.Get(ctx, draftExternalCloudAuditBackendKey)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	out, err := services.UnmarshalExternalCloudAudit(item.Value)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return out, nil
+}
+
+// UpsertDraftExternalAudit upserts the draft external cloud audit resource.
+func (s *ExternalCloudAuditService) UpsertDraftExternalCloudAudit(ctx context.Context, in *externalcloudaudit.ExternalCloudAudit) (*externalcloudaudit.ExternalCloudAudit, error) {
+	err := backend.RunWhileLocked(ctx, backend.RunWhileLockedConfig{
+		LockConfiguration: backend.LockConfiguration{
+			Backend:  s.backend,
+			LockName: externalCloudAuditLockName,
+			TTL:      externalCloudAuditLockTTL,
+		},
+	}, func(ctx context.Context) error {
+		value, err := services.MarshalExternalCloudAudit(in)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		_, err = s.backend.Put(ctx, backend.Item{
+			Key:   draftExternalCloudAuditBackendKey,
+			Value: value,
+		})
+		return trace.Wrap(err)
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return in, nil
+}
+
+// DeleteDraftExternalAudit removes the draft external cloud audit resource.
+func (s *ExternalCloudAuditService) DeleteDraftExternalCloudAudit(ctx context.Context) error {
```

Comment:
Is there a check somewhere that what we're actually deleting is indeed a draft, and not a live configuration? The PRD mentions adding a `external_audit_config`, and I'm a bit worried that although the CLI tool distinguishes between deleting a draft and a live configuration, the service might be called by something else.

(That, of course, if manipulating the draft doesn't require the same privileges; otherwise, a potential attack vector would be to delete the live configuration anyway. Still, it sounds like making sure we don't delete a `MetaNameExternalCloudAuditCluster` this way a good sanity check from the defensive programming standpoint.)

Reply:
I am not sure if I understand. We got 2 singleton properties stored under different backend key. 
`Draft` is deleted when calling `DeleteDraftExternalCloudAudit`
`Cluster` is deleted when calling `DisableClusterExternalCloudAudit`.

We have UT covering it.

The API exposed by service is simple:
```
// ExternalCloudAuditGetter defines an interface for reading external cloud audits.
type ExternalCloudAuditGetter interface {
	// GetDraftExternalCloudAudit returns the draft external cloud audit resource.
	GetDraftExternalCloudAudit(context.Context) (*externalcloudaudit.ExternalCloudAudit, error)
	// GetClusterExternalCloudAudit returns the cluster external cloud audit resource.
	GetClusterExternalCloudAudit(context.Context) (*externalcloudaudit.ExternalCloudAudit, error)
}

// ExternalCloudAudit defines an interface for managing ExternalCloudAudit.
type ExternalCloudAudit interface {
	ExternalCloudAuditGetter

	// UpsertDraftExternalCloudAudit upserts the draft external cloud audit resource.
	UpsertDraftExternalCloudAudit(context.Context, *externalcloudaudit.ExternalCloudAudit) (*externalcloudaudit.ExternalCloudAudit, error)
	// DeleteDraftExternalCloudAudit deletes the draft external cloud audit resource.
	DeleteDraftExternalCloudAudit(context.Context) error
	// PromoteToClusterExternalCloudAudit promotes draft to cluster external
	// cloud audit resource.
	PromoteToClusterExternalCloudAudit(context.Context) error
	// DisableClusterExternalCloudAudit disables an cluster external cloud audit
	// resource.
	DisableClusterExternalCloudAudit(context.Context) error
}
```
you can see that both DeleteDraft and DisableCluster does not accept any parameters (beside context) so it's hard to misuse it.

But maybe I am missing your point.

Reply:
I see. Never mind, my mistake. Thanks!